### DISCUSSION TITLE
16 bit wide bus in SaxonSoC

### DISCUSSION THREAD

**Discussion by surabibio** [SpinalHDL/SpinalHDL#660] [Q&A]
I am trying to connect a 16 bit APB3 bus to the SaxonSoc. I am building over the Ulx3sMinimal.scala file.
 This is what I have done:

    implicit val interconnect = BmbInterconnectGenerator()
    implicit val bmbPeripheral = BmbBridgeGenerator(mapping = SizeMapping(0x10000000, 16 MiB)).peripheral(dataWidth = 32)
    implicit val peripheralDecoder = bmbPeripheral.asPeripheralDecoder() //Will be used by peripherals as default bus to connect to
    implicit val cpu = VexRiscvBmbGenerator()

    val dBus16 = BmbBridgeGenerator()
    dBus16.dataWidth(16)
    
    val dpr = BmbOnChipRamGenerator(0x90000000L)
    dpr.size.load(1 KiB)
    dpr.hexInit.loadNothing()
    
    val bridge16 = BmbBridgeGenerator(mapping = SizeMapping(0xa0000000L, 1 MiB)).peripheral(dataWidth = 16)
      bridge16.dataWidth(16)
    
    val decoder16  =  bridge16.asPeripheralDecoder()
    
    val apb3Mapping : SizeMapping = (0x10000000L, 64 KiB)
    val apb3 = BmbToApb3Generator(mapping = apb3Mapping) (interconnect, decoder16  )
    
      apb3.apb3Config load Apb3Config(
        addressWidth = log2Up(apb3Mapping.size),
        dataWidth = 16,
        useSlaveError = false
      )
    
    
    //Interconnect specification
    interconnect.addConnection(
      cpu.iBus    -> List(ramA.ctrl),
      cpu.dBus    -> List(ramA.ctrl, bmbPeripheral.bmb, dBus16.bmb), 
      dBus16.bmb  -> List(dpr.ctrl, bridge16.bmb)
    )

```

But I keep getting a error at BmbToApb3Bridge.scala:23 which is 
  assert(apb3Config.dataWidth == bmbParameter.access.dataWidth)


What am I missing?
``

**Comment by Dolu1990** [comment]
Your code seems fine, maybe there is some hardcoded 32 bits in the SaxonSoc libraries.

Do you have your code somewere for me to give a try ?

**Reply by surabibio** [reply]
Hi Charles,
Thankyou for your attention.
When I make my BmbToApb3Generator dataWidth to 16 bit it gives out the error.

I am using dev-03 Saxon Repository.
[code.zip](https://github.com/SpinalHDL/SpinalHDL/files/8333124/code.zip)

**Reply by Dolu1990** [reply]
Hi,

Can you send me the broken version ? to be sure i'm in the same case as you.

Thanks ^^

**Reply by surabibio** [reply]
[Ulx3sMinimal.zip](https://github.com/SpinalHDL/SpinalHDL/files/8349224/Ulx3sMinimal.zip)
Hi Charles,
This is the 16 bit APB3 bus that gives the error.
